datafusion/src/logical_plan/builder.rs

```diff
@@ -1090,6 +1094,101 @@ pub(crate) fn expand_wildcard(
     }
 }
 
+/// check whether the logical plan we are building is valid
+fn check_plan_invalid(plan: &LogicalPlan) -> Option<DataFusionError> {
+    match plan {
+        LogicalPlan::Projection(Projection { expr, input, .. })
+        | LogicalPlan::Sort(Sort { expr, input })
+        | LogicalPlan::Window(Window {
+            window_expr: expr,
+            input,
+            ..
+        }) => check_plan_invalid(input)
+            .or_else(|| check_any_invalid_expr(expr, input.schema())),
+
+        LogicalPlan::Filter(Filter {
+            predicate: expr,
+            input,
+        }) => {
+            check_plan_invalid(input).or_else(|| check_invalid_expr(expr, input.schema()))
+        }
+
+        LogicalPlan::Aggregate(Aggregate {
+            input,
+            group_expr,
+            aggr_expr,
+            ..
+        }) => {
+            let schema = input.schema();
+            check_plan_invalid(input)
+                .or_else(|| check_any_invalid_expr(group_expr, schema))
+                .or_else(|| check_any_invalid_expr(aggr_expr, schema))
+        }
+
+        LogicalPlan::Join(Join { left, right, .. })
+        | LogicalPlan::CrossJoin(CrossJoin { left, right, .. }) => {
+            check_plan_invalid(left).or_else(|| check_plan_invalid(right))
+        }
+
+        LogicalPlan::Repartition(Repartition { input, .. })
+        | LogicalPlan::Limit(Limit { input, .. })
+        | LogicalPlan::Explain(Explain { plan: input, .. })
+        | LogicalPlan::Analyze(Analyze { input, .. }) => check_plan_invalid(input),
+
+        LogicalPlan::Union(Union { inputs, .. }) => {
+            inputs.iter().filter_map(check_plan_invalid).next()
+        }
+
+        LogicalPlan::TableScan(TableScan {
+            table_name: _,
+            source,
+            projection,
+            projected_schema,
+            filters,
+            limit: _,
+        }) => {
+            if let Some(projection) = projection {
+                if projection.len() > projected_schema.fields().len() {
+                    return Some(DataFusionError::Plan(
+                        "projection contains columns that doesnt belong to projected schema"
+                            .to_owned(),
+                    ));
+                }
+            }
+            let schema = &source.schema().to_dfschema_ref().ok()?;
+            check_any_invalid_expr(filters, schema)
+        }
+
+        _ => None,
+    }
+}
+
+/// find first error in the exprs
+fn check_any_invalid_expr(
+    exprs: &[Expr],
+    schema: &DFSchemaRef,
+) -> Option<DataFusionError> {
+    exprs
+        .iter()
+        .filter_map(|e| check_invalid_expr(e, schema))
+        .next()
+}
```

Comment:
```suggestion
/// check whether the logical plan we are building is valid
fn check_plan_invalid(plan: &LogicalPlan) -> Result<()> {
    match plan {
        LogicalPlan::Projection(Projection { expr, input, .. })
        | LogicalPlan::Sort(Sort { expr, input })
        | LogicalPlan::Window(Window {
            window_expr: expr,
            input,
            ..
        }) => check_plan_invalid(input)
            .or_else(|_| check_any_invalid_expr(expr, input.schema())),

        LogicalPlan::Filter(Filter {
            predicate: expr,
            input,
        }) => check_plan_invalid(input)
            .or_else(|_| check_invalid_expr(expr, input.schema())),

        LogicalPlan::Aggregate(Aggregate {
            input,
            group_expr,
            aggr_expr,
            ..
        }) => {
            let schema = input.schema();
            check_plan_invalid(input)
                .or_else(|_| check_any_invalid_expr(group_expr, schema))
                .or_else(|_| check_any_invalid_expr(aggr_expr, schema))
        }

        LogicalPlan::Join(Join { left, right, .. })
        | LogicalPlan::CrossJoin(CrossJoin { left, right, .. }) => {
            check_plan_invalid(left).or_else(|_| check_plan_invalid(right))
        }

        LogicalPlan::Repartition(Repartition { input, .. })
        | LogicalPlan::Limit(Limit { input, .. })
        | LogicalPlan::Explain(Explain { plan: input, .. })
        | LogicalPlan::Analyze(Analyze { input, .. }) => check_plan_invalid(input),

        LogicalPlan::Union(Union { inputs, .. }) => {
            inputs.iter().map(check_plan_invalid).collect()
        }

        LogicalPlan::TableScan(TableScan {
            table_name: _,
            source,
            projection,
            projected_schema,
            filters,
            limit: _,
        }) => {
            if let Some(projection) = projection {
                if projection.len() > projected_schema.fields().len() {
                    return Err(DataFusionError::Plan(
                        "projection contains columns that doesnt belong to projected schema"
                            .to_owned(),
                    ));
                }
            }
            let schema = &source.schema().to_dfschema_ref()?;
            check_any_invalid_expr(filters, schema)
        }

        _ => Ok(()),
    }
}
```

Reply:
I'm sorry for the late response! Thanks @jackwener ! 
The code looks better! I'm going to apply your suggestions, but need resolve some errors like:
``check_plan_invalid(input)
            .or_else(|_| check_any_invalid_expr(expr, input.schema()))``
We should change `or_else` to `and_then`.